flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamConfigChainer.java

```diff
@@ -121,6 +124,91 @@ private void head(OperatorID headOperatorID) {
 			TypeSerializer<IN> inputSerializer,
 			TypeSerializer<OUT> outputSerializer,
 			boolean createKeyedStateBackend) {
+		return chain(operatorID,
+			operatorFactory,
+			inputSerializer,
+			outputSerializer,
+			createKeyedStateBackend,
+			null);
+	}
+
+	public <T> StreamConfigChainer<OWNER> chainWithNonChainedOutputs(
+		OperatorID operatorID,
+		OneInputStreamOperator<T, T> operator,
+		TypeSerializer<T> typeSerializer,
+		boolean createKeyedStateBackend,
+		List<StreamEdge> nonChainedOutputs) {
+		return chainWithNonChainedOutputs(operatorID, operator, typeSerializer, typeSerializer, createKeyedStateBackend, nonChainedOutputs);
+	}
+
+	public <T> StreamConfigChainer<OWNER> chainWithNonChainedOutputs(
+		OneInputStreamOperator<T, T> operator,
+		TypeSerializer<T> typeSerializer,
+		List<StreamEdge> nonChainedOutputs) {
+		return chainWithNonChainedOutputs(new OperatorID(), operator, typeSerializer, nonChainedOutputs);
+	}
+
+	public <T> StreamConfigChainer<OWNER> chainWithNonChainedOutputs(
+		OperatorID operatorID,
+		OneInputStreamOperator<T, T> operator,
+		TypeSerializer<T> typeSerializer,
+		List<StreamEdge> nonChainedOutputs) {
+		return chainWithNonChainedOutputs(operatorID, operator, typeSerializer, typeSerializer, false, nonChainedOutputs);
+	}
+
+	public <T> StreamConfigChainer<OWNER> chainWithNonChainedOutputs(
+		OneInputStreamOperatorFactory<T, T> operatorFactory,
+		TypeSerializer<T> typeSerializer,
+		List<StreamEdge> nonChainedOutputs) {
+		return chainWithNonChainedOutputs(new OperatorID(), operatorFactory, typeSerializer, nonChainedOutputs);
+	}
+
+	public <T> StreamConfigChainer<OWNER> chainWithNonChainedOutputs(
+		OperatorID operatorID,
+		OneInputStreamOperatorFactory<T, T> operatorFactory,
+		TypeSerializer<T> typeSerializer,
+		List<StreamEdge> nonChainedOutputs) {
+		return chainWithNonChainedOutputs(operatorID, operatorFactory, typeSerializer, typeSerializer, false, nonChainedOutputs);
+	}
+
+	private <IN, OUT> StreamConfigChainer<OWNER> chainWithNonChainedOutputs(
```

Comment:
I think this builder class is growing a bit too big with too many overloaded methods. I think we need to split it into two step builder pattern. 

Instead of providing many overloaded variants of `chain` and `chainWithNonChainedOutputs` (and maybe more in the future), I think it would be better to have a single `chain` method with just the obligatory parameters
```			
StreamConfigEdgeChainer chain(TypeSerializer<T> typeSerializer);
```
(?)
that returns `StreamConfigEdgeChainer` (different name?). `StreamConfigEdgeChainer` would be the 2nd level builder allowing to set more optional parameters like operator factory, nonChainedOutputs, createKeyedStateBackend, .....

After calling `StreamConfigEdgeChainer.build()` it would return again the first level builder `StreamConfigChainer`.


```
testHarness
  .chain(typeSerializer1)
    .setOperatorID(...)
    .setOutputTypeSerializer(...)
    .build()
  .chain(typeSerializer2)
    .setNonChainedOutputsCount(2)
    .build()
  .finish();
```
WDYT?

(for the sake of backward compatbility the existing `chain` methods could be left as they are (`@Deprecated`), so that we don't have to modify all of the existing tests.